Mathlib/AlgebraicGeometry/EllipticCurve/NormalForms.lean

```diff
@@ -0,0 +1,667 @@
+  a `WeierstrassCurve` is of form $Y^2 + XY = X^3 + a_2X^2 + a_6$ or
+  $Y^2 + a_3Y = X^3 + a_4X + a_6$.
+  It is the normal form of characteristic = 2.
+
+  If the field is of characteristic = 2, then there exists a change of variables which will change
+  it into such normal form (`WeierstrassCurve.exists_variableChange_isCharTwoNF`).
+  See also `WeierstrassCurve.toCharTwoNF` and
+  `WeierstrassCurve.toCharTwoNF_spec`.
+
+## References
+
+* [J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, weierstrass equation, normal form
+
+-/
+
+variable {R : Type*} [CommRing R] (W : WeierstrassCurve R)
+variable {F : Type*} [Field F] (E : EllipticCurve F)
+
+namespace WeierstrassCurve
+
+/-! ### Normal forms of characteristic ≠ 2 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic ≠ 2, if its $a_1, a_3 = 0$.
+In other words it is $Y^2 = X^3 + a_2X^2 + a_4X + a_6$. -/
+@[mk_iff]
+structure IsCharNeTwoNF : Prop where
+  a₁ : W.a₁ = 0
+  a₃ : W.a₃ = 0
+
+namespace IsCharNeTwoNF
+
+variable {W} (self : W.IsCharNeTwoNF)
+include self
+
+theorem b₂ : W.b₂ = 4 * W.a₂ := by
+  simp_rw [WeierstrassCurve.b₂, self.a₁]
+  ring1
+
+theorem b₄ : W.b₄ = 2 * W.a₄ := by
+  simp_rw [WeierstrassCurve.b₄, self.a₃]
+  ring1
+
+theorem b₆ : W.b₆ = 4 * W.a₆ := by
+  simp_rw [WeierstrassCurve.b₆, self.a₃]
+  ring1
+
+theorem b₈ : W.b₈ = 4 * W.a₂ * W.a₆ - W.a₄ ^ 2 := by
+  simp_rw [WeierstrassCurve.b₈, self.a₁, self.a₃]
+  ring1
+
+theorem c₄ : W.c₄ = 16 * W.a₂ ^ 2 - 48 * W.a₄ := by
+  simp_rw [WeierstrassCurve.c₄, self.b₂, self.b₄]
+  ring1
+
+theorem c₆ : W.c₆ = -64 * W.a₂ ^ 3 + 288 * W.a₂ * W.a₄ - 864 * W.a₆ := by
+  simp_rw [WeierstrassCurve.c₆, self.b₂, self.b₄, self.b₆]
+  ring1
+
+theorem Δ : W.Δ = -64 * W.a₂ ^ 3 * W.a₆ + 16 * W.a₂ ^ 2 * W.a₄ ^ 2 - 64 * W.a₄ ^ 3
+    - 432 * W.a₆ ^ 2 + 288 * W.a₂ * W.a₄ * W.a₆ := by
+  simp_rw [WeierstrassCurve.Δ, self.b₂, self.b₄, self.b₆, self.b₈]
+  ring1
+
+end IsCharNeTwoNF
+
+section VariableChange
+
+variable [Invertible (2 : R)]
+
+/-- This is an explicit change of variables of a `WeierstrassCurve` to
+a normal form of characteristic ≠ 2, provided that 2 is invertible in the ring. -/
+@[simps]
+def toCharNeTwoNF : VariableChange R where
+  u := 1
+  r := 0
+  s := ⅟2 * -W.a₁
+  t := ⅟2 * -W.a₃
+
+theorem toCharNeTwoNF_spec : (W.variableChange W.toCharNeTwoNF).IsCharNeTwoNF := by
+  constructor <;> simp [← mul_assoc]
+
+theorem exists_variableChange_isCharNeTwoNF :
+    ∃ C : VariableChange R, (W.variableChange C).IsCharNeTwoNF :=
+  ⟨_, W.toCharNeTwoNF_spec⟩
+
+end VariableChange
+
+/-! ### Normal forms of characteristic ≠ 2 or 3 -/
+
+/-- A `WeierstrassCurve` is in normal form of characteristic ≠ 2 or 3, if its $a_1, a_2, a_3 = 0$.
+In other words it is $Y^2 = X^3 + a_4X + a_6$.
+
+This is also the normal form of characteristic = 3 and j = 0. -/
+@[mk_iff]
+structure IsCharNeTwoThreeNF : Prop where
```

Comment:
I think it's better to call this `IsCharThreeJZeroNF` to match `IsCharThreeJNeZeroNF` (analogous to the difference between `IsCharTwoJZeroNF` and `IsCharTwoJNeZeroNF`).

Personally I find the names slightly confusing, because this is the normal form for both characteristic not 3 and characteristic 3 but j = 0. Arguably you could give them your own names e.g. `IsNF(B|C)(1|2)` based on Silverman's numbering in Appendix A, or `IsNF(A.)*` etc based on the parameters killed off. I don't know if this is a good idea, but I also don't have any better suggestions, so feel free to ignore this.

Reply:
I think `IsShortNF` is good. But I'm not sure about `IsCharNeTwoNF -> IsMediumNF`.

Reply:
I'm happy with `IsShortNF` and we can keep `IsCharNeTwoNF` too. I guess `IsShort` is outrageous? ;)

Reply:
`IsShort` -> Time is short 🤣